apollo-router/src/spec/selection.rs

```diff
@@ -113,41 +110,12 @@ impl Selection {
                     .unwrap_or(current_type)
                     .to_owned();
 
-                let fragment_type = &type_condition;
-
-                // this is the type we pass when extracting the fragment's selections
-                // If the type condition is a union or interface and the current type implements it, then we want
-                // to keep the current type when extracting the fragment's selections, as it is more precise
-                // than the interface.
-                // If it is not, then we use the type condition
-                let relevant_type = if schema.is_interface(type_condition.as_str()) {
-                    // Query validation should have already verified that current type implements that interface
-                    debug_assert!(
-                        schema.is_subtype(
-                            type_condition.as_str(),
-                            current_type
-                        ) || schema.is_implementation(
-                            type_condition.as_str(),
-                            current_type
-                        )
-                     ||
-                        // if the current type and the type condition are both the same interface, it is still valid
-                        type_condition.as_str()
-                            == current_type
-                    );
-                    let relevant_type = schema.most_precise(current_type, fragment_type);
-                    debug_assert!(relevant_type.is_some());
-                    relevant_type.unwrap_or(fragment_type)
-                } else {
-                    fragment_type
-                };
-
                 let selection_set = inline_fragment
                     .selection_set
                     .selections
                     .iter()
                     .filter_map(|selection| {
-                        Selection::from_hir(selection, relevant_type, schema, count, defer_stats)
+                        Selection::from_hir(selection, type_condition.as_str(), count, defer_stats)
```

Comment:
that can't be right. The code we had previously was designed to avoid using the type condition if we already knew a more precise type. How is this avoiding that bug?

Reply:
sorry it's still a draft while i prepare a test

Reply:
fwiw the tests are all fully passing, and we did write tests for the type_condition bug we saw, so unless there's a failing test here I don't see it being an issue

Reply:
Why is it important or useful to know a more precise type?

---

This PR is not just a chore/refactor, it’s a fix for the bug reported in https://github.com/apollographql/router/issues/4214. The validity rule for fragment spreads or inline fragments is:

https://spec.graphql.org/October2021/#sec-Fragment-spread-is-possible
> * Let _applicableTypes_ be the intersection of [GetPossibleTypes](https://spec.graphql.org/October2021/#GetPossibleTypes())(_fragmentType_) and [GetPossibleTypes](https://spec.graphql.org/October2021/#GetPossibleTypes())(_parentType_)
> * _applicableTypes_ must not be empty.

There are two relevant sets of types, but neither is necessarily a subset of the other. They only need to have some intersection. The logic of this `debug_assert` and of `fn most_precise` makes incorrect assumptions about what’s valid.